feat: clients common trait impl

Otter-0x4ka5h opened the pull request:
Things added extra here:
- CPI Return Type (Return)
   Clone + Copy: It's a lightweight wrapper around PhantomData

- CPI Accounts (AccountInfo wrappers)
   Clone add because devs might call the same program multiple times with the same accounts

- Copy for Bumps.

Closes #3857

Commit:
feat: clients common trait impl

fix: conflicts

Commit:
fix: debug error on unimplemented traits

0x4ka5h commented on lang/syn/src/codegen/program/instruction.rs:
```diff
@@ -25,6 +89,20 @@ pub fn generate(program: &Program) -> proc_macro2::TokenStream {
+                || ix
+                    .args
+                    .iter()
+                    .all(|arg| can_derive_common_trait(&arg.raw_arg.ty));
```
sure, we can do that

0x4ka5h commented on lang/syn/src/codegen/program/instruction.rs:
```diff
@@ -3,6 +3,70 @@ use crate::parser;
+                return false;
+            }
+            let first_segment = &segments[0];
+            let ident_str = first_segment.ident.to_string();
```
ahh yeah, it only check first segment and breaks full paths rights, i think we should check last segments which extract soo inner type?

Commit:
requested updates

0x4ka5h commented on lang/syn/src/codegen/program/instruction.rs:
```diff
@@ -25,6 +89,20 @@ pub fn generate(program: &Program) -> proc_macro2::TokenStream {
                 })
                 .collect();
+
+            // Check if all argument types can derive Clone and Debug
```
The existing `can_derive_clone_ty` and `can_derive_debug_ty` functions in `common.rs` work with `IdlType`, while this code works with `syn::Type`. These are different type systems, so those functions don't apply here.

I feel the current changes make sense.

Commit:
fix(fmt): cargo fmt on lang

jamie-osec committed:
fix: address review feedback on common-trait derives

- Restore unconditional Debug + Borsh-gated Clone for declare_program! type
  definitions; the can_derive_*_ty helpers wrongly returned false for generic
  IdlTypes, which would have stripped Debug/Clone from generic IDL types.
- Fix the broken test_convert_idl_type_to_str: the function now returns
  Result<String, syn::Error>, so the test must unwrap.
- Replace the trailing-comma quote!(Clone, Debug,) trick in instruction.rs
  with a Vec<TokenStream> spliced via #(, #extra_derives)*.
- Extend common.rs tests with generic-struct, oversize-array, and wrapper
  fixtures covering negative cases for clone/debug/default/copy.
- Add tests for the new can_derive_common_trait helper in instruction.rs
  covering primitives, fully-qualified paths, Option/Vec recursion, arrays,
  tuples, references, user-defined types, and exotic types.

jamie-osec committed:
chore: clippy clean-up for common-trait derives

- Annotate the new can_derive_clone_ty / can_derive_debug_ty `Defined`
  arms with `#[allow(clippy::expect_used, reason = ...)]`, matching the
  existing pattern on can_derive_copy_ty / can_derive_default_ty. The
  `.expect("Type def must exist")` is sound: IDL cross-references are
  guaranteed consistent by Anchor tooling.
- Replace the manual `is_empty + unwrap()` segment lookup in
  can_derive_common_trait with a `let-else` so the unwrap goes away.
- Allow `indexing_slicing` and `unwrap_used` at the test module level
  for declare_program::common; fixtures have known shape and inputs
  are valid by construction.

jamie-osec committed:
docs: Add CHANGELOG entry for client common-trait derives
